src/build/streams.ts

```diff
@@ -46,3 +76,153 @@ export function pipeStreams(streams: PipeStream[]): NodeJS.ReadableStream {
         return a.pipe(b);
       });
 }
+
+/**
+ * An asynchronous queue that is read as an async iterable.
+ */
+class AsyncQueue<V> implements AsyncIterable<V> {
+  private blockedOn: Deferred<IteratorResult<V>>|undefined = undefined;
+  backlog: Array<{value: IteratorResult<V>, deferred: Deferred<void>}> = [];
+  private _closed = false;
+
+  /**
+   * Add the given value onto the queue.
+   *
+   * The return value of this method resolves once the value has been removed
+   * from the queue. Useful for flow control.
+   *
+   * Must not be called after the queue has been closed.
+   */
+  async write(value: V) {
+    if (this._closed) {
+      throw new Error('Wrote to closed writable iterable');
+    }
+    return this._write({value, done: false});
+  }
+
+  /**
+   * Close the queue, indicating that no more values will be written.
+   *
+   * If this method is not called, a consumer iterating over the values will
+   * wait forever.
+   *
+   * The returned promise resolves once the consumer has been notified of the
+   * end of the queue.
+   */
+  async close() {
+    this._closed = true;
+    return this._write({done: true} as any);
+  }
+
+  private async _write(value: IteratorResult<V>) {
+    if (this.blockedOn) {
+      this.blockedOn.resolve(value);
+      this.blockedOn = undefined;
+    } else {
+      const deferred = new Deferred<void>();
+      this.backlog.push({value, deferred});
+      await deferred.promise;
+    }
+  }
+
+  /**
+   * Iterate over values in the queue. Not intended for multiple readers.
```

Comment:
Not sure the part about multiple readers is needed. Iterators hold state and are never intended for multiple readers. Normal use is that readers receive Iterables, which produce a new Iterator for each reader.